File: src/hud_mod.py
"""HUD centering for Crimson Desert ultrawide monitors.

Modifies UI HTML/CSS in PAZ archive 0012 to constrain HUD elements
within a 21:9 safe area, centering them on ultrawide displays.
"""

import os
import re
import sys
from pathlib import Path

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from paz_crypto import decrypt, encrypt, lz4_decompress, lz4_compress
from paz_parse import parse_pamt
from paz_repack import _match_compressed_size, _save_timestamps

_UI_ARCHIVE = '0012'

_HUD_HTML_PATHS = (
    'ui/minimaphudview2.html',
    'ui/statusgaugeview2.html',
)
_CSS_PATH = 'ui/gamecommon.css'
_ALL_PATHS = (*_HUD_HTML_PATHS, _CSS_PATH)

_SAFEFRAME_ID = 'HUDSafeFrame'
_SAFEFRAME_CLASS = 'ui-view-max-size-21-9'
_SAFEFRAME_MAX_WIDTH = 2520
_SAFEFRAME_OPEN = f'    <div id="{_SAFEFRAME_ID}" class="{_SAFEFRAME_CLASS}">'
_SAFEFRAME_CLOSE = '    </div>'

_BODY_OPEN_RE = re.compile(r'(<body\b[^>]*>)', re.IGNORECASE | re.DOTALL)
_BODY_CLOSE_RE = re.compile(r'(</body>)', re.IGNORECASE)
_HTML_COMMENT_RE = re.compile(r'<!--.*?-->', re.DOTALL)

_GAUGE_TRACKER = '<div id="HPGaugeTracker"'
_GAUGE_SKILLPOINT = '<div id="UIHudScaleSkillPointStatusGaugeContainer"'


# ── Text compaction (matching centerhud approach) ──────────────────

def _compact_css(text):
    t = re.sub(r'\n\s*\n+', '\n', text)
    t = re.sub(r'[ \t]+\n', '\n', t)
    return re.sub(r'\s*([{}:;,])\s*', r'\1', t)


def _compact_text(entry_path, text):
    if entry_path.lower().endswith('.css'):
        return _compact_css(text)
    t = text
    if entry_path.lower().endswith('.html'):
        t = _HTML_COMMENT_RE.sub('', t)
    return re.sub(r'\n\s*\n+', '\n', t)


# ── PAZ I/O ────────────────────────────────────────────────────────

def _find_ui_entries(game_dir):
    pamt = os.path.join(game_dir, _UI_ARCHIVE, '0.pamt')
    paz_dir = os.path.join(game_dir, _UI_ARCHIVE)
    if not os.path.exists(pamt):
        raise RuntimeError(f'UI archive not found: {pamt}')
    entries = parse_pamt(pamt, paz_dir=paz_dir)
    by_path = {e.path.lower(): e for e in entries}
    result = []
    for p in _ALL_PATHS:
        e = by_path.get(p)
        if e is None:
            raise RuntimeError(f'{p} not found in archive')
        result.append(e)
    return result


def _read_raw(entry):
    with open(entry.paz_file, 'rb') as f:
        f.seek(entry.offset)
        return f.read(entry.comp_size)


def _decode_text(entry, raw):
    """Decode payload, detecting encryption. Returns (text, encrypted)."""
    if entry.compressed:
        if entry.compression_type != 2:
            raise RuntimeError(f'Unsupported compression for {entry.path}')
        try:
            plain = lz4_decompress(raw, entry.orig_size)
        except Exception:
            dec = decrypt(raw, Path(entry.path).name)
            plain = lz4_decompress(dec, entry.orig_size)
            return plain.rstrip(b'\x00').decode('utf-8', errors='replace'), True
        try:
            return plain.rstrip(b'\x00').decode('utf-8'), False
        except UnicodeDecodeError:
            pass
        try:
            dec = decrypt(raw, Path(entry.path).name)
            dp = lz4_decompress(dec, entry.orig_size)
            return dp.rstrip(b'\x00').decode('utf-8', errors='replace'), True
        except Exception:
            return plain.rstrip(b'\x00').decode('utf-8', errors='replace'), False
    try:
        return raw.rstrip(b'\x00').decode('utf-8'), False
    except UnicodeDecodeError:
        dec = decrypt(raw, Path(entry.path).name)
        return dec.rstrip(b'\x00').decode('utf-8', errors='replace'), True


# ── HTML modification ──────────────────────────────────────────────

def _modify_html(entry_path, text):
    if _SAFEFRAME_ID in text:
        return text
    if not _BODY_OPEN_RE.search(text):
        raise RuntimeError(f'No <body> in {entry_path}')
    if not _BODY_CLOSE_RE.search(text):
        raise RuntimeError(f'No </body> in {entry_path}')

    modified = _BODY_OPEN_RE.sub(
        rf'\1\n{_SAFEFRAME_OPEN}\n', text, count=1)

    if entry_path.lower() == 'ui/statusgaugeview2.html':
        if _GAUGE_TRACKER not in modified:
            raise RuntimeError('HPGaugeTracker not found')
        if _GAUGE_SKILLPOINT not in modified:
            raise RuntimeError('SkillPointContainer not found')
        modified = modified.replace(
            _GAUGE_TRACKER,
            f'{_SAFEFRAME_CLOSE}\n{_GAUGE_TRACKER}', 1)
        modified = modified.replace(
            _GAUGE_SKILLPOINT,
            f'{_SAFEFRAME_OPEN}\n{_GAUGE_SKILLPOINT}', 1)

    modified = _BODY_CLOSE_RE.sub(
        rf'{_SAFEFRAME_CLOSE}\n\1', modified, count=1)
    return modified


# ── CSS modification ───────────────────────────────────────────────

def _modify_css(text, max_width=_SAFEFRAME_MAX_WIDTH):
    """Update .ui-view-max-size-21-9 max-width in existing rule."""
    pat = re.compile(
        r'(\.ui-view-max-size-21-9\s*\{)([^}]*)(\})',
        re.IGNORECASE | re.DOTALL)
    m = pat.search(text)
    if m is None:
        raise RuntimeError('.ui-view-max-size-21-9 not found in CSS')

    body = m.group(2)
    prop_re = re.compile(r'(max-width\s*:\s*)([^;]+)(;)', re.IGNORECASE)
    if prop_re.search(body):
        new_body = prop_re.sub(
            rf'\g<1>{max_width}px\3', body, count=1)
    else:
        new_body = body.rstrip()
        if new_body and not new_body.endswith(';'):
            new_body += ';'
        new_body += f' max-width: {max_width}px;'

    return text[:m.start(2)] + new_body + text[m.end(2):]


# ── Payload building (follows centerhud pattern) ───────────────────

def _build_payload(modified_text, entry, encrypted):
    compacted = _compact_text(entry.path, modified_text)
    candidates = [compacted] if compacted != modified_text else []
    candidates.append(modified_text)

    last_err = None
    for candidate in candidates:
        plaintext = candidate.encode('utf-8')
        try:
            if entry.compressed:
                adjusted = _match_compressed_size(
                    plaintext, entry.comp_size, entry.orig_size)
                payload = lz4_compress(adjusted)
                if len(payload) != entry.comp_size:
                    raise RuntimeError(
                        f'comp_size mismatch: {len(payload)} != {entry.comp_size}')
            else:
                if len(plaintext) > entry.comp_size:
                    raise RuntimeError(
                        f'Too large: {len(plaintext)} > {entry.comp_size}')
                payload = plaintext + b'\x00' * (entry.comp_size - len(plaintext))
        except Exception as e:
            last_err = e
            continue

        if encrypted:
            payload = encrypt(payload, Path(entry.path).name)
            if len(payload) != entry.comp_size:
                raise RuntimeError('Encrypted size mismatch')
        return payload

    raise RuntimeError(f'Cannot fit {entry.path}: {last_err}')


# ── Backup management ─────────────────────────────────────────────

def _backups_dir():
    return os.path.join(
        os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
        'backups', 'hud')


def _backup_exists():
    return os.path.exists(os.path.join(_backups_dir(), 'meta.txt'))


def _save_backups(entries):
    bdir = _backups_dir()
    os.makedirs(bdir, exist_ok=True)
    for entry in entries:
        raw = _read_raw(entry)
        fname = entry.path.replace('/', '__') + '.bin'
        with open(os.path.join(bdir, fname), 'wb') as f:
            f.write(raw)
    meta = '\n'.join(
        f'{e.path}|{e.comp_size}|{e.orig_size}|{e.offset}'
        for e in entries)
    with open(os.path.join(bdir, 'meta.txt'), 'w') as f:
        f.write(meta)


def _load_backup(entry):
    fname = entry.path.replace('/', '__') + '.bin'
    path = os.path.join(_backups_dir(), fname)
    if not os.path.exists(path):
        return None
    with open(path, 'rb') as f:
        return f.read()


# ── Public API ─────────────────────────────────────────────────────

def install_centered_hud(game_dir, max_width=_SAFEFRAME_MAX_WIDTH):
    print(f'  [HUD] Finding UI entries (max-width: {max_width}px)...')
    entries = _find_ui_entries(game_dir)

    decoded = {}
    for entry in entries:
        raw = _read_raw(entry)
        text, enc = _decode_text(entry, raw)
        decoded[entry.path] = (text, enc)

    html_already = all(
        _SAFEFRAME_ID in decoded[e.path][0]
        for e in entries if e.path.lower().endswith('.html'))
    if html_already:
        print('  [HUD] Already installed.')
        return {'status': 'ok'}

    if not _backup_exists():
        print('  [HUD] Saving backups...')
        _save_backups(entries)

    writes = []
    for entry in entries:
        text, enc = decoded[entry.path]
        if entry.path.lower().endswith('.html'):
            modified = _modify_html(entry.path, text)
        elif entry.path.lower().endswith('.css'):
            modified = _modify_css(text, max_width=max_width)
        else:
            continue
        if modified == text:
            continue
        payload = _build_payload(modified, entry, enc)
        writes.append((entry, payload))

    for entry, payload in writes:
        restore_ts = _save_timestamps(entry.paz_file)
        with open(entry.paz_file, 'r+b') as f:
            f.seek(entry.offset)
            f.write(payload)
        restore_ts()
        print(f'  [HUD] Patched: {entry.path}')

    print('  [HUD] Done!')
    return {'status': 'ok'}


def restore_hud(game_dir):
    if not _backup_exists():
        return {'status': 'no_backup'}
    entries = _find_ui_entries(game_dir)

    for entry in entries:
        data = _load_backup(entry)
        if data is None:
            return {'status': 'error'}
        if len(data) != entry.comp_size:
            return {'status': 'stale_backup'}

    for entry in entries:
        data = _load_backup(entry)
        restore_ts = _save_timestamps(entry.paz_file)
        with open(entry.paz_file, 'r+b') as f:
            f.seek(entry.offset)
            f.write(data)
        restore_ts()
        print(f'  [HUD] Restored: {entry.path}')

    return {'status': 'ok'}

File: src/UltimateCameraMod/Services/HudMod.cs
using System.Text;
using System.Text.RegularExpressions;
using UltimateCameraMod.Paz;

namespace UltimateCameraMod.Services;

/// <summary>
/// HUD centering for ultrawide monitors. Updates UI HTML/CSS in archive 0012.
/// </summary>
public static class HudMod
{
    private const string UiArchive = "0012";
    private static readonly string[] HudHtmlPaths = { "ui/minimaphudview2.html", "ui/statusgaugeview2.html" };
    private const string CssPath = "ui/gamecommon.css";

    private const string SafeFrameId = "HUDSafeFrame";
    private const int DefaultMaxWidth = 1920;

    // CentreHUD approach: 16:9 uses class "ui-view-max-size-16-9", 21:9 uses "ui-view-max-size-21-9"
    private static string GetSafeFrameClass(int maxWidth) => maxWidth > 1920 ? "ui-view-max-size-21-9" : "ui-view-max-size-16-9";
    private static string GetSafeFrameOpen(int maxWidth) => $"    <div id=\"{SafeFrameId}\" class=\"{GetSafeFrameClass(maxWidth)}\">";
    private const string SafeFrameClose = "    </div>";

    private static readonly Regex BodyOpenRe = new(@"(<body\b[^>]*>)", RegexOptions.IgnoreCase | RegexOptions.Singleline | RegexOptions.Compiled);
    private static readonly Regex BodyCloseRe = new(@"(</body>)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
    private static readonly Regex HtmlCommentRe = new(@"<!--.*?-->", RegexOptions.Singleline | RegexOptions.Compiled);

    private const string GaugeTracker = "<div id=\"HPGaugeTracker\"";
    private const string GaugeSkillPoint = "<div id=\"UIHudScaleSkillPointStatusGaugeContainer\"";

    public static Func<string>? BackupsDirOverride { get; set; }
    private static string BackupsDir => BackupsDirOverride?.Invoke()
        ?? Path.Combine(AppContext.BaseDirectory, "backups", "hud");

    // ── Text compaction ──────────────────────────────────────────────

    private static string CompactCss(string text)
    {
        var t = Regex.Replace(text, @"\n\s*\n+", "\n");
        t = Regex.Replace(t, @"[ \t]+\n", "\n");
        return Regex.Replace(t, @"\s*([{}:;,])\s*", "$1");
    }

    private static string CompactText(string entryPath, string text)
    {
        if (entryPath.EndsWith(".css", StringComparison.OrdinalIgnoreCase))
            return CompactCss(text);
        var t = text;
        if (entryPath.EndsWith(".html", StringComparison.OrdinalIgnoreCase))
            t = HtmlCommentRe.Replace(t, "");
        return Regex.Replace(t, @"\n\s*\n+", "\n");
    }

    // ── PAZ I/O ──────────────────────────────────────────────────────

    private static List<PazEntry> FindUiEntries(string gameDir)
    {
        string pamt = Path.Combine(gameDir, UiArchive, "0.pamt");
        string pazDir = Path.Combine(gameDir, UiArchive);
        if (!File.Exists(pamt))
            throw new FileNotFoundException($"UI archive not found: {pamt}");

        var entries = PamtReader.Parse(pamt, pazDir);
        var byPath = new Dictionary<string, PazEntry>();
        foreach (var e in entries)
        {
            string key = e.Path.ToLowerInvariant();
            if (!byPath.ContainsKey(key))
                byPath[key] = e;
        }

        var allPaths = HudHtmlPaths.Append(CssPath);
        var result = new List<PazEntry>();
        foreach (var p in allPaths)
        {
            if (!byPath.TryGetValue(p, out var entry))
                throw new FileNotFoundException($"{p} not found in archive");
            result.Add(entry);
        }
        return result;
    }

    private static byte[] ReadRaw(PazEntry entry)
    {
        using var fs = new FileStream(entry.PazFile, FileMode.Open, FileAccess.Read);
        fs.Seek(entry.Offset, SeekOrigin.Begin);
        byte[] data = new byte[entry.CompSize];
        int totalRead = 0;
        while (totalRead < data.Length)
        {
            int n = fs.Read(data, totalRead, data.Length - totalRead);
            if (n == 0) throw new EndOfStreamException();
            totalRead += n;
        }
        return data;
    }

    private static (string Text, bool Encrypted) DecodeText(PazEntry entry, byte[] raw)
    {
        if (entry.Compressed)
        {
            if (entry.CompressionType != 2)
                throw new InvalidOperationException($"Unsupported compression for {entry.Path}");
            try
            {
                byte[] plain = CompressionUtils.Lz4Decompress(raw, entry.OrigSize);
                try { return (Encoding.UTF8.GetString(plain).TrimEnd('\0'), false); }
                catch (DecoderFallbackException) { }
            }
            catch { }

            try
            {
                byte[] dec = AssetCodec.Decode(raw, Path.GetFileName(entry.Path));
                byte[] dp = CompressionUtils.Lz4Decompress(dec, entry.OrigSize);
                return (Encoding.UTF8.GetString(dp).TrimEnd('\0'), true);
            }
            catch { }

            byte[] fallback = CompressionUtils.Lz4Decompress(raw, entry.OrigSize);
            return (Encoding.UTF8.GetString(fallback).TrimEnd('\0'), false);
        }

        try { return (Encoding.UTF8.GetString(raw).TrimEnd('\0'), false); }
        catch
        {
            byte[] dec = AssetCodec.Decode(raw, Path.GetFileName(entry.Path));
            return (Encoding.UTF8.GetString(dec).TrimEnd('\0'), true);
        }
    }

    // ── HTML modification ────────────────────────────────────────────

    private static string ModifyHtml(string entryPath, string text, int maxWidth)
    {
        string sfOpen = GetSafeFrameOpen(maxWidth);
        if (text.Contains(SafeFrameId)) return text;
        if (!BodyOpenRe.IsMatch(text)) throw new InvalidOperationException($"No <body> in {entryPath}");
        if (!BodyCloseRe.IsMatch(text)) throw new InvalidOperationException($"No </body> in {entryPath}");

        string modified = BodyOpenRe.Replace(text, $"$1\n{sfOpen}\n", 1);

        if (entryPath.Equals("ui/statusgaugeview2.html", StringComparison.OrdinalIgnoreCase))
        {
            if (!modified.Contains(GaugeTracker)) throw new InvalidOperationException("HPGaugeTracker not found");
            if (!modified.Contains(GaugeSkillPoint)) throw new InvalidOperationException("SkillPointContainer not found");
            modified = modified.Replace(GaugeTracker, $"{SafeFrameClose}\n{GaugeTracker}");
            modified = modified.Replace(GaugeSkillPoint, $"{sfOpen}\n{GaugeSkillPoint}");
        }

        modified = BodyCloseRe.Replace(modified, $"{SafeFrameClose}\n$1", 1);
        return modified;
    }

    // ── CSS modification ─────────────────────────────────────────────

    private static string ModifyCss(string text, int maxWidth = DefaultMaxWidth)
    {
        // CentreHUD approach: update 21-9 to 2520px, then create 16-9 with user's width
        var pat219 = new Regex(@"(\.ui-view-max-size-21-9\s*\{)([^}]*)(\})", RegexOptions.IgnoreCase | RegexOptions.Singleline);
        var m219 = pat219.Match(text);
        if (!m219.Success) throw new InvalidOperationException(".ui-view-max-size-21-9 not found in CSS");

        // Step 1: ensure 21-9 has max-width: 2520px
        string body219 = m219.Groups[2].Value;
        var propRe = new Regex(@"(max-width\s*:\s*)([^;]+)(;)", RegexOptions.IgnoreCase);
        string newBody219;
        if (propRe.IsMatch(body219))
            newBody219 = propRe.Replace(body219, "${1}2520px${3}", 1);
        else
            newBody219 = body219.TrimEnd() + " max-width: 2520px;";
        string modified = text[..m219.Groups[2].Index] + newBody219 + text[(m219.Groups[2].Index + m219.Groups[2].Length)..];

        // Step 2: check if 16-9 rule exists
        var pat169 = new Regex(@"(\.ui-view-max-size-16-9\s*\{)([^}]*)(\})", RegexOptions.IgnoreCase | RegexOptions.Singleline);
        var m169 = pat169.Match(modified);
        if (m169.Success)
        {
            // Update existing 16-9 rule
            string body169 = m169.Groups[2].Value;
            string newBody169;
            if (propRe.IsMatch(body169))
                newBody169 = propRe.Replace(body169, $"${{1}}{maxWidth}px${{3}}", 1);
            else
                newBody169 = body169.TrimEnd() + $" max-width: {maxWidth}px;";
            return modified[..m169.Groups[2].Index] + newBody169 + modified[(m169.Groups[2].Index + m169.Groups[2].Length)..];
        }

        // Step 3: create 16-9 by cloning 21-9 properties with different max-width
        var m219new = pat219.Match(modified);
        string clonedBody = m219new.Groups[2].Value;
        clonedBody = propRe.Replace(clonedBody, $"${{1}}{maxWidth}px${{3}}", 1);
        string newRule = $".ui-view-max-size-16-9 {{{clonedBody}}}";
        int insertAt = m219new.Index + m219new.Length;
        return modified[..insertAt] + "\n" + newRule + modified[insertAt..];
    }

    // ── Payload building ─────────────────────────────────────────────

    private static byte[] BuildPayload(string modifiedText, PazEntry entry, bool encrypted)
    {
        string compacted = CompactText(entry.Path, modifiedText);
        var candidates = new List<string>();
        if (compacted != modifiedText) candidates.Add(compacted);
        candidates.Add(modifiedText);

        Exception? lastErr = null;
        foreach (var candidate in candidates)
        {
            byte[] plaintext = Encoding.UTF8.GetBytes(candidate);
            try
            {
                byte[] payload;
                if (entry.Compressed)
                {
                    byte[] adjusted = entry.IsHtml
                        ? ArchiveWriter.MatchCompressedSizeHtml(plaintext, entry.CompSize, entry.OrigSize)
                        : entry.IsCss
                            ? ArchiveWriter.MatchCompressedSizeCss(plaintext, entry.CompSize, entry.OrigSize)
                            : ArchiveWriter.MatchCompressedSize(plaintext, entry.CompSize, entry.OrigSize);
                    payload = CompressionUtils.Lz4Compress(adjusted);
                    if (payload.Length != entry.CompSize)
                        throw new InvalidOperationException($"comp_size mismatch: {payload.Length} != {entry.CompSize}");
                }
                else
                {
                    if (plaintext.Length > entry.CompSize)
                        throw new InvalidOperationException($"Too large: {plaintext.Length} > {entry.CompSize}");
                    payload = new byte[entry.CompSize];
                    Array.Copy(plaintext, payload, plaintext.Length);
                }

                if (encrypted)
                {
                    payload = AssetCodec.Encode(payload, Path.GetFileName(entry.Path));
                    if (payload.Length != entry.CompSize)
                        throw new InvalidOperationException("Encrypted size mismatch");
                }
                return payload;
            }
            catch (Exception e) { lastErr = e; }
        }

        throw new InvalidOperationException($"Cannot fit {entry.Path}: {lastErr?.Message}");
    }

    // ── Backup management ────────────────────────────────────────────

    private static bool BackupExists() => File.Exists(Path.Combine(BackupsDir, "meta.txt"));

    private static void SaveBackups(List<PazEntry> entries)
    {
        Directory.CreateDirectory(BackupsDir);
        foreach (var entry in entries)
        {
            byte[] raw = ReadRaw(entry);
            string fname = entry.Path.Replace('/', '_').Replace('\\', '_') + ".bin";
            File.WriteAllBytes(Path.Combine(BackupsDir, fname), raw);
        }
        var meta = string.Join('\n', entries.Select(e => $"{e.Path}|{e.CompSize}|{e.OrigSize}|{e.Offset}"));
        File.WriteAllText(Path.Combine(BackupsDir, "meta.txt"), meta);
    }

    private static byte[]? LoadBackup(PazEntry entry)
    {
        string fname = entry.Path.Replace('/', '_').Replace('\\', '_') + ".bin";
        string path = Path.Combine(BackupsDir, fname);
        return File.Exists(path) ? File.ReadAllBytes(path) : null;
    }

    // ── Live detection ─────────────────────────────────────────────

    public static bool DetectHudModified(string gameDir)
    {
        try
        {
            var entries = FindUiEntries(gameDir);
            foreach (var entry in entries)
            {
                if (!entry.Path.EndsWith(".html", StringComparison.OrdinalIgnoreCase))
                    continue;
                byte[] raw = ReadRaw(entry);
                var (text, _) = DecodeText(entry, raw);
                if (text.Contains(SafeFrameId))
                    return true;
            }
        }
        catch { }
        return false;
    }

    // ── Public API ───────────────────────────────────────────────────

    public static Dictionary<string, object> InstallCenteredHud(string gameDir,
        int maxWidth = DefaultMaxWidth, Action<string>? log = null)
    {
        log?.Invoke($"[HUD] Finding UI entries (max-width: {maxWidth}px)...");
        var entries = FindUiEntries(gameDir);

        var decoded = new Dictionary<string, (string Text, bool Encrypted)>();
        foreach (var entry in entries)
        {
            byte[] raw = ReadRaw(entry);
            decoded[entry.Path] = DecodeText(entry, raw);
        }

        bool alreadyInstalled = entries
            .Where(e => e.Path.EndsWith(".html", StringComparison.OrdinalIgnoreCase))
            .Any(e => decoded[e.Path].Text.Contains(SafeFrameId));

        if (!BackupExists())
        {
            log?.Invoke("[HUD] Saving backups...");
            SaveBackups(entries);
        }

        // If already installed (possibly different mode), restore vanilla first then reinstall
        if (alreadyInstalled)
        {
            log?.Invoke("[HUD] Restoring vanilla before reinstall...");
            foreach (var entry in entries)
            {
                byte[]? backup = LoadBackup(entry);
                if (backup == null || backup.Length != entry.CompSize) continue;
                var restoreTs = ArchiveWriter.SaveTimestamps(entry.PazFile);
                using (var fs = new FileStream(entry.PazFile, FileMode.Open, FileAccess.Write))
                {
                    fs.Seek(entry.Offset, SeekOrigin.Begin);
                    fs.Write(backup);
                }
                restoreTs();
            }
            // Re-decode from restored vanilla
            decoded.Clear();
            foreach (var entry in entries)
            {
                byte[] raw = ReadRaw(entry);
                decoded[entry.Path] = DecodeText(entry, raw);
            }
        }

        var writes = new List<(PazEntry Entry, byte[] Payload)>();
        foreach (var entry in entries)
        {
            var (text, enc) = decoded[entry.Path];
            string modified;
            if (entry.Path.EndsWith(".html", StringComparison.OrdinalIgnoreCase))
                modified = ModifyHtml(entry.Path, text, maxWidth);
            else if (entry.Path.EndsWith(".css", StringComparison.OrdinalIgnoreCase))
                modified = ModifyCss(text, maxWidth);
            else continue;

            if (modified == text) continue;
            byte[] payload = BuildPayload(modified, entry, enc);
            writes.Add((entry, payload));
        }

        foreach (var (entry, payload) in writes)
        {
            var restoreTs = ArchiveWriter.SaveTimestamps(entry.PazFile);
            using (var fs = new FileStream(entry.PazFile, FileMode.Open, FileAccess.Write))
            {
                fs.Seek(entry.Offset, SeekOrigin.Begin);
                fs.Write(payload);
            }
            restoreTs();
            log?.Invoke($"[HUD] Updated: {entry.Path}");
        }

        log?.Invoke("[HUD] Done!");
        return new() { ["status"] = "ok" };
    }

    public static Dictionary<string, object> RestoreHud(string gameDir, Action<string>? log = null)
    {
        if (!BackupExists())
            return new() { ["status"] = "no_backup" };

        var entries = FindUiEntries(gameDir);

        foreach (var entry in entries)
        {
            byte[]? data = LoadBackup(entry);
            if (data == null) return new() { ["status"] = "error" };
            if (data.Length != entry.CompSize) return new() { ["status"] = "stale_backup" };
        }

        foreach (var entry in entries)
        {
            byte[] data = LoadBackup(entry)!;
            var restoreTs = ArchiveWriter.SaveTimestamps(entry.PazFile);
            using (var fs = new FileStream(entry.PazFile, FileMode.Open, FileAccess.Write))
            {
                fs.Seek(entry.Offset, SeekOrigin.Begin);
                fs.Write(data);
            }
            restoreTs();
            log?.Invoke($"[HUD] Restored: {entry.Path}");
        }

        return new() { ["status"] = "ok" };
    }
}
